Content.Server/Storage/EntitySystems/EntityStorageSystem.cs

```diff
@@ -59,12 +114,239 @@ public void EmptyContents(EntityUid uid, EntityStorageComponent? component = nul
+        {
+            if (_container.IsEntityInContainer(entity))
+                continue;
+
+            if (!ev.ContentsWhitelist.Contains(entity))
+                if (!CanFit(entity, uid))
+                    continue;
+            
+            if (!AddToContents(entity, uid, component))
+                continue;
+
+            count++;
+            if (count >= component.StorageCapacityMax)
+                break;
+        }
+
+        ModifyComponents(uid, component);
+        SoundSystem.Play(component.CloseSound.GetSound(), Filter.Pvs(uid), uid);
+        component.LastInternalOpenAttempt = default;
+    }
+
+    public bool Insert(EntityUid toInsert, EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        if (component.Open)
+        {
+            Transform(toInsert).WorldPosition = Transform(container).WorldPosition;
+            return true;
+        }
+
+        return component.Contents.Insert(toInsert, EntityManager);
+    }
+
+    public bool Remove(EntityUid toRemove, EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        return component.Contents.Remove(toRemove, EntityManager);
+    }
+
+    public bool CanInsert(EntityUid container, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(container, ref component))
+            return false;
+
+        if (component.Open)
+            return true;
+
+        if (component.Contents.ContainedEntities.Count >= component.StorageCapacityMax)
+            return false;
+
+        return true;
+    }
+
+    public bool TryOpenStorage(EntityUid user, EntityUid target)
+    {
+        if (!CanOpen(user, target))
+            return false;
+
+        OpenStorage(target);
+        return true;
+    }
+
+    public bool TryCloseStorage(EntityUid target)
+    {
+        if (!CanClose(target))
+        {
+            return false;
+        }
+
+        CloseStorage(target);
+        return true;
+    }
+
+    public bool CanOpen(EntityUid user, EntityUid target, bool silent = false, EntityStorageComponent? component = null)
+    {
+        if (!Resolve(target, ref component))
+            return false;
+
+        if (component.IsWeldedShut)
+        {
+            if (!silent && !component.Contents.Contains(user))
+                _popupSystem.PopupEntity(Loc.GetString("entity-storage-component-welded-shut-message"), target, Filter.Pvs(target));
+
+            return false;
+        }
+
+        if (TryComp<LockComponent>(target, out var lockComp) && lockComp.Locked)
+        {
+            if (!silent)
+                _popupSystem.PopupEntity(Loc.GetString("entity-storage-component-locked-message"), target, Filter.Pvs(target));
+
+            return false;
+        }
+
+        var ev = new StorageOpenAttemptEvent();
+        RaiseLocalEvent(target, ev, true);
```

Comment:
The lock system should just subscribe to this event. Same goes for the crematorium.